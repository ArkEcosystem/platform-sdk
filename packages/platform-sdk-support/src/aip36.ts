import Joi, { ValidationError } from "joi";
import { join } from "path";

import { Business } from "./aip36/business";
import { Delegate } from "./aip36/delegate";
import { Media } from "./aip36/media";
import { Meta } from "./aip36/meta";
import { Module } from "./aip36/module";
import { Plugin } from "./aip36/plugin";
import { Product } from "./aip36/product";
import { SocialMedia } from "./aip36/social-media";
import { SourceControl } from "./aip36/source-control";

export class AIP36 {
	#data: Record<string, unknown> = {};

	public business(): Business {
		return new Business(this.#data);
	}

	public delegate(): Delegate {
		return new Delegate(this.#data);
	}

	public media(): Media {
		return new Media(this.#data);
	}

	public meta(): Meta {
		return new Meta(this.#data);
	}

	public module(): Module {
		return new Module(this.#data);
	}

	public plugin(): Plugin {
		return new Plugin(this.#data);
	}

	public product(): Product {
		return new Product(this.#data);
	}

	public socialMedia(): SocialMedia {
		return new SocialMedia(this.#data);
	}

	public sourceControl(): SourceControl {
		return new SourceControl(this.#data);
	}

	public validate(): void {
		const { error, value } = Joi.object({
			business: Joi.object({
				type: Joi.string().required(),
			}),
			delegate: Joi.object({
				type: Joi.string().valid("public", "private").required(),
				payout: Joi.object({
					frequency: Joi.object({
						type: Joi.string().valid("day", "week", "month", "quarter", "year").required(),
						value: Joi.number().min(1).max(365).required(),
					}),
					percentage: Joi.object({
						min: Joi.number().min(1).max(100).required(),
						max: Joi.number().min(1).max(100).required(),
					}),
				}),
			}),
			images: Joi.array().items(
				Joi.object({
					type: Joi.valid("logo", "image"),
					value: Joi.alternatives().try(
						Joi.string().pattern(/(?:https?:)?\/\/(?:i\.)?(?:imgur\.com)\/([A-z0-9]*)(\.[A-z]{3,4})/),
						Joi.string().pattern(
							/(?:https?:)?\/(?:raw\.githubusercontent\.com)\/([A-Za-z0-9-_.]+)(?:\/[A-z0-9_-].*)(\.(jpe?g|png|gif))$/,
						),
						Joi.string().pattern(
							/(?:https?:)?\/(?:gitlab\.com)\/([A-Za-z0-9-_.].*)(?:\/-\/raw)(?:\/[A-z0-9_-].*)(\.(jpe?g|png|gif))$/,
						),
						Joi.string().pattern(
							/(?:https?:)?\/\/(?:\w+\.)(?:staticflickr\.com)\/([A-z0-9-_]+)(?:\/[A-z0-9_-]+)(\.[A-z]{3,4})/,
						),
					),
				}),
			),
			meta: Joi.object({
				description: Joi.string(),
				displayName: Joi.string(),
				website: Joi.string().uri(),
			}),
			module: Joi.object({
				developedBy: Joi.string().required(),
				network: Joi.string().required(),
				releaseDate: Joi.date().iso().required(),
				platform: Joi.string().required(),
				requirements: Joi.array().items(Joi.string()).required(),
			}),
			plugin: Joi.object({
				developedBy: Joi.string().required(),
				network: Joi.string().required(),
				releaseDate: Joi.date().iso().required(),
				platform: Joi.string().required(),
				requirements: Joi.array().items(Joi.string()).required(),
			}),
			product: Joi.object({
				developedBy: Joi.string().required(),
				network: Joi.string().required(),
				releaseDate: Joi.date().iso().required(),
				platform: Joi.string().required(),
			}),
			socialMedia: Joi.array().items(
				Joi.object({
					type: Joi.valid(
						"discord",
						"facebook",
						"instagram",
						"linkedin",
						"medium",
						"reddit",
						"slack",
						"telegram",
						"twitter",
						"youtube",
					).required(),
					value: Joi.alternatives()
						.try(
							Joi.string().pattern(
								/(?:https?:)?\/\/(?:www\.)?(?:discord\.com\/invite|discord.gg(\/invite)?)\/([A-z0-9]+)/,
							),
							Joi.string().pattern(/(?:https?:)?\/\/(?:www\.)?(?:facebook\.com)\/([A-Za-z0-9-_.]+)/),
							Joi.string().pattern(/(?:https?:)?\/\/(?:www\.)?(?:instagram\.com)\/([A-Za-z0-9-_.]+)/),
							Joi.string().pattern(
								/(?:https?:)?\/\/(?:www\.)?(?:linkedin\.com)\/(?:in|company)\/([\w\-_À-ÿ%]+)\/?/,
							),
							Joi.string().pattern(/(?:https?:)?\/\/(?:www\.)?(?:medium\.com)\/([A-Za-z0-9-_.@]+)/),
							Joi.string().pattern(
								/(?:https?:)?\/\/(?:www\.)?(?:reddit\.com)\/(?:u(?:ser)?|r)\/([A-z0-9-_]+)\/?/,
							),
							Joi.string().pattern(/(?:https?:)?\/\/(?:[a-z]+)\.(?:slack\.com)/),
							Joi.string().pattern(
								/(?:https?:)?\/\/(?:t(?:elegram)?\.me|telegram\.org)\/([a-z0-9_]{5,32})/,
							),
							Joi.string().pattern(/(?:https?:)?\/\/(?:www\.)?(?:twitter\.com)\/@?([A-z0-9_]+)/),
							Joi.string().pattern(
								/(?:https?:)?\/\/(?:www\.)?(?:youtube\.com)\/(?:channel|user)\/([A-z0-9]+)\/?/,
							),
						)
						.required(),
				}),
			),
			sourceControl: Joi.array().items(
				Joi.object({
					type: Joi.valid("bitbucket", "github", "gitlab", "npm").required(),
					value: Joi.alternatives()
						.try(
							Joi.string().pattern(
								/(?:https?:)?\/\/(?:www\.)?(?:bitbucket\.org)\/([A-Za-z0-9-_.]+)(?:\/[A-z0-9_-]+)?\/?/,
							),
							Joi.string().pattern(
								/(?:https?:)?\/\/(?:www\.)?(?:github\.com)\/([A-Za-z0-9-_.]+)(?:\/[A-z0-9_-]+)?\/?/,
							),
							Joi.string().pattern(
								/(?:https?:)?\/\/(?:www\.)?(?:gitlab\.com)\/([A-Za-z0-9-_.]+)(?:\/[A-z0-9_-]+)?\/?/,
							),
							Joi.string().pattern(
								/(?:https?:)?\/\/(?:www\.)?(?:npmjs\.com)\/package\/([A-Za-z0-9-_.@]+)/,
							),
						)
						.required(),
				}),
			),
			videos: Joi.array().items(
				Joi.object({
					type: Joi.valid("video").required(),
					value: Joi.alternatives()
						.try(
							Joi.string().pattern(
								/(?:https?:)?\/\/(?:www\.)?(?:vimeo\.com|player\.vimeo\.com\/video)\/([0-9]+)/,
							),
							Joi.string().pattern(
								/(?:https?:)?\/\/(?:(?:www\.)?(?:youtube\.com)\/(?:watch\?v=|embed\/)|youtu\.be\/)([A-z0-9-_]+)\/?/,
							),
						)
						.required(),
				}),
			),
		}).validate(this.#data);

		if (error instanceof ValidationError) {
			throw error;
		}

		this.#data = value;
	}

	public toString(): string {
		return JSON.stringify(this.#data);
	}
}
