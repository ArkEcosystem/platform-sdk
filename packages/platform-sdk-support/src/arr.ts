export class Arr {
	public static randomElement(items: any[]) {
		return items[Math.floor(Math.random() * items.length)];
	}
}
