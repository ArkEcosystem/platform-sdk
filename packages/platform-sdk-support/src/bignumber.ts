import BigNumberJS from "bignumber.js";

export type NumberLike = string | number | BigNumberJS | BigNumber;

const BigNumberClone = BigNumberJS.clone({ DECIMAL_PLACES: 8, EXPONENTIAL_AT: 1e9 });

// This implementation is significantly slower than the native BigInt but for
// applications that use the Platform SDK this performance loss is acceptable.
export class BigNumber {
	public static readonly ZERO: BigNumber = new BigNumber(0);
	public static readonly ONE: BigNumber = new BigNumber(1);
	public static readonly SATOSHI: BigNumber = new BigNumber(1e8);

	readonly #value: BigNumberJS;

	private constructor(value: NumberLike, decimals?: number) {
		this.#value = this.toBigNumber(value);

		if (decimals !== undefined) {
			this.#value = this.#value.decimalPlaces(decimals);
		}
	}

	public static make(value: NumberLike, decimals?: number): BigNumber {
		return new BigNumber(value, decimals);
	}

	public decimalPlaces(decimals?: number): BigNumber {
		return BigNumber.make(this.#value, decimals);
	}

	public plus(value: NumberLike): BigNumber {
		return BigNumber.make(this.#value.plus(this.toBigNumber(value)));
	}

	public minus(value: NumberLike): BigNumber {
		return BigNumber.make(this.#value.minus(this.toBigNumber(value)));
	}

	public divide(value: NumberLike): BigNumber {
		return BigNumber.make(this.#value.dividedBy(this.toBigNumber(value)));
	}

	public times(value: NumberLike): BigNumber {
		return BigNumber.make(this.#value.multipliedBy(this.toBigNumber(value)));
	}

	public isNaN(): boolean {
		return this.#value.isNaN();
	}

	public isPositive(): boolean {
		return this.#value.isPositive();
	}

	public isFinite(): boolean {
		return this.#value.isFinite();
	}

	public comparedTo(value: NumberLike): number {
		return this.#value.comparedTo(this.toBigNumber(value));
	}

	public isEqualTo(value: NumberLike): boolean {
		return this.#value.isEqualTo(this.toBigNumber(value));
	}

	public isGreaterThan(value: NumberLike): boolean {
		return this.#value.isGreaterThan(this.toBigNumber(value));
	}

	public isGreaterThanOrEqualTo(value: NumberLike): boolean {
		return this.#value.isGreaterThanOrEqualTo(this.toBigNumber(value));
	}

	public isLessThan(value: NumberLike): boolean {
		return this.#value.isLessThan(this.toBigNumber(value));
	}

	public isLessThanOrEqualTo(value: NumberLike): boolean {
		return this.#value.isLessThanOrEqualTo(this.toBigNumber(value));
	}

	public toSatoshi(): BigNumber {
		return BigNumber.make(this.#value.multipliedBy(1e8));
	}

	public toFixed(decimals?: number): string {
		return decimals ? this.#value.toFixed(decimals) : this.#value.toFixed();
	}

	public toNumber(): number {
		return this.#value.toNumber();
	}

	public toString(): string {
		return this.#value.toString();
	}

	public valueOf(): string {
		return this.#value.valueOf();
	}

	private toBigNumber(value: NumberLike): BigNumberJS {
		if (value instanceof BigNumber) {
			return new BigNumberClone(value.valueOf());
		}

		return new BigNumberClone(value);
	}
}
