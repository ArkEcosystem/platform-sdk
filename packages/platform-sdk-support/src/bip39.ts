export class BIP39 {
	public static normalize(passphrase: string): string {
		return passphrase.normalize("NFD");
	}
}
