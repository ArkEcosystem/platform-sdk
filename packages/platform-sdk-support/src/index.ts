export * from "./censor";
export * from "./arr";
export * from "./bignumber";
export * from "./bip39";
export * from "./bip44";
export * from "./buffoon";
export * from "./currency-formatter";
export * from "./http";
