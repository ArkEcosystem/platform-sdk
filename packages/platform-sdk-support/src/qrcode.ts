import QRious from "qrious";

export class QRCode {
	readonly #instance: QRious;

	private constructor() {
		this.#instance = new QRious();
	}

	public static fromString(value: string): QRCode {
		return new QRCode().set({ value });
	}

	public static fromObject(value: object): QRCode {
		return new QRCode().set({
			value: JSON.stringify(value),
		});
	}

	public set(value: object): QRCode {
		this.#instance.set(value);

		return this;
	}

	public toDataURL(): string {
		return this.#instance.toDataURL("image/png");
	}
}
