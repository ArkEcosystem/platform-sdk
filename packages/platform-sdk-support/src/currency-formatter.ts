import { BigNumber, NumberLike } from "./bignumber";

export class CurrencyFormatter {
	// todo: implement generic formatting method
	// https://github.com/ArkEcosystem/desktop-wallet/blob/develop/src/renderer/mixins/currency.js#L7-L100

	public static simpleFormatCrypto(value: NumberLike, token: string): string {
		return `${value} ${token}`;
	}

	public static toBuilder(value: NumberLike, decimals = 8): BigNumber {
		return BigNumber.make(value, decimals);
	}

	public static subToUnit(value: NumberLike, decimals = 8): BigNumber {
		return BigNumber.make(value, decimals).divide(1e8);
	}

	public static unitToSub(value: NumberLike, decimals = 8): BigNumber {
		return BigNumber.make(value, decimals).toSatoshi();
	}

	public static cryptoToCurrency(
		value: NumberLike,
		price: NumberLike,
		options: { fromSubUnit: boolean; decimals: number } = {
			fromSubUnit: true,
			decimals: 2,
		},
	): string {
		return (options.fromSubUnit ? this.subToUnit(value) : BigNumber.make(value))
			.decimalPlaces(options.decimals)
			.times(price)
			.toFixed();
	}
}
