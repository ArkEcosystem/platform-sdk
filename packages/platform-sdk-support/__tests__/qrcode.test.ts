import "jest-extended";

import { QRCode } from "../src/qrcode";

test("#fromString", () => {
	expect(QRCode.fromString("https://google.com")).toBeInstanceOf(QRCode);
});

test("#fromObject", () => {
	expect(QRCode.fromObject({ url: "https://google.com" })).toBeInstanceOf(QRCode);
});

test("#toDataURL", async () => {
	const actual: string = await QRCode.fromString("https://google.com").toDataURL();

	expect(actual).toStartWith("data:image/png;base64,");
	expect(actual).toMatchSnapshot();
});

describe.each(["utf8", "svg", "terminal"])("%s", (type) => {
	it("should turn into a string", async () => {
		await expect(QRCode.fromString("https://google.com").toString(type as any)).resolves.toMatchSnapshot();
	});
});
