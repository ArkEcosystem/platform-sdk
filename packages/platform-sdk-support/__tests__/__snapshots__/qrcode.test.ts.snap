// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`#toDataURL 1`] = `"data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAIQAAACECAYAAABRRIOnAAAAAklEQVR4AewaftIAAAOKSURBVO3BQY5jiRUDwcwH3f/KdC+84OoDglTlmTYjzB/M/NcxU46ZcsyUY6YcM+WYKcdMOWbKMVOOmXLMlGOmHDPlmCnHTHnxIZXflIRPqLwjCU9UWhKaym9KwieOmXLMlGOmvPiyJHyTyjtU3pGEJyrflIRvUvmmY6YcM+WYKS9+mMo7kvAOlSdJaCqfSEJT+YTKO5Lwk46ZcsyUY6a8+Mskoak8UWlJaEl4koS/yTFTjplyzJQXf7kkNJUnKi0JT1RaEv7NjplyzJRjprz4YUn4TSrvSEJTaSo/KQn/JMdMOWbKMVNefJnK/1ISmkpLQlNpSWgqLQlN5R0q/2THTDlmyjFTzB/8i6l8IglNpSXh/8kxU46ZcswU8wcfUGlJaCrflITfpPIkCU9UvikJP+mYKcdMOWaK+YMvUnmShE+otCQ0lSdJ+CaVloQnKi0J71BpSfimY6YcM+WYKS9+mUpLwhOVloSm8ptUWhKaSkvCE5UnSXii0pLwiWOmHDPlmCkvviwJTaUloak8ScI3qTxJQlNpSWgqn0jCO5LQVL7pmCnHTDlmivmDD6g8SUJT+UQSmkpLQlNpSfhJKk+S0FRaEprKO5LwiWOmHDPlmCkvviwJTaUloak8SUJTaUn4hMo7kvAkCU9UWhKaSktCU2lJ+KZjphwz5ZgpL35YEj6h0pLQVD6RhKbyROUdSXii8gmVloRPHDPlmCnHTHnxy1RaEppKS0JTaUl4ovJE5UkSmspPSsKTJDSVbzpmyjFTjpny4oeptCQ0lScqLQlN5ZuS0FQ+ofIJlZaEloRvOmbKMVOOmWL+4F9M5R1J+ITKkyS8Q+VJEn7TMVOOmXLMlBcfUvlNSWhJeIfKkyQ8SUJTeaLSkvAkCU3lHUn4xDFTjplyzJQXX5aEb1J5otKS8I4kPFFpSXhHEt6h0pLwm46ZcsyUY6a8+GEq70jCP0kSmsoTlU8koam0JDSVloRPHDPlmCnHTHnxl1NpSWgqn0hCU3mShKbSVJ6o/KRjphwz5ZgpL/4yKi0JT5LQVJ4koam0JLwjCU3lSRKayjcdM+WYKcdMefHDkvCTkvAJlSdJaCpPVL4pCb/pmCnHTDlmyosvU/lNKu9IwpMkNJWWhCcqLQlN5R0qv+mYKcdMOWaK+YOZ/zpmyjFTjplyzJRjphwz5Zgpx0w5ZsoxU46ZcsyUY6YcM+WYKf8B7SOFJl3T5SkAAAAASUVORK5CYII="`;

exports[`svg should turn into a string 1`] = `
"<svg xmlns=\\"http://www.w3.org/2000/svg\\" viewBox=\\"0 0 33 33\\" shape-rendering=\\"crispEdges\\"><path fill=\\"#ffffff\\" d=\\"M0 0h33v33H0z\\"/><path stroke=\\"#000000\\" d=\\"M4 4.5h7m4 0h3m2 0h1m1 0h7M4 5.5h1m5 0h1m2 0h3m2 0h3m1 0h1m5 0h1M4 6.5h1m1 0h3m1 0h1m1 0h2m1 0h1m2 0h1m1 0h1m1 0h1m1 0h3m1 0h1M4 7.5h1m1 0h3m1 0h1m1 0h1m1 0h3m5 0h1m1 0h3m1 0h1M4 8.5h1m1 0h3m1 0h1m1 0h1m2 0h3m2 0h1m1 0h1m1 0h3m1 0h1M4 9.5h1m5 0h1m1 0h4m1 0h4m1 0h1m5 0h1M4 10.5h7m1 0h1m1 0h1m1 0h1m1 0h1m1 0h1m1 0h7M12 11.5h4m1 0h1M4 12.5h1m1 0h5m7 0h2m2 0h5M5 13.5h2m4 0h1m1 0h2m5 0h1m2 0h1m3 0h1M5 14.5h3m2 0h1m1 0h3m4 0h1m1 0h1m1 0h3m1 0h2M6 15.5h1m1 0h2m2 0h5m1 0h1m1 0h2m1 0h1m4 0h1M4 16.5h2m1 0h4m1 0h1m1 0h1m6 0h2m1 0h1m1 0h3M4 17.5h1m1 0h1m1 0h2m1 0h1m2 0h1m1 0h3m4 0h1m1 0h1m1 0h1M4 18.5h1m3 0h3m1 0h3m2 0h5m1 0h3m1 0h2M4 19.5h1m2 0h1m1 0h1m1 0h1m2 0h1m1 0h2m1 0h3m1 0h2m3 0h1M4 20.5h1m1 0h1m1 0h3m1 0h3m1 0h1m1 0h7m1 0h1M12 21.5h3m4 0h2m3 0h2M4 22.5h7m4 0h2m3 0h1m1 0h1m1 0h1m1 0h3M4 23.5h1m5 0h1m1 0h1m2 0h1m2 0h1m1 0h1m3 0h2M4 24.5h1m1 0h3m1 0h1m1 0h1m2 0h1m3 0h6m1 0h1m1 0h1M4 25.5h1m1 0h3m1 0h1m1 0h1m1 0h1m1 0h4m1 0h2m1 0h5M4 26.5h1m1 0h3m1 0h1m1 0h2m3 0h2m1 0h1m4 0h2m1 0h1M4 27.5h1m5 0h1m5 0h2m1 0h7m2 0h1M4 28.5h7m1 0h3m1 0h1m1 0h1m2 0h1m1 0h6\\"/></svg>
"
`;

exports[`terminal should turn into a string 1`] = `
"[47m  [0m[47m  [0m[47m  [0m[47m  [0m[47m  [0m[47m  [0m[47m  [0m[47m  [0m[47m  [0m[47m  [0m[47m  [0m[47m  [0m[47m  [0m[47m  [0m[47m  [0m[47m  [0m[47m  [0m[47m  [0m[47m  [0m[47m  [0m[47m  [0m[47m  [0m[47m  [0m[47m  [0m[47m  [0m[47m  [0m[47m  [0m
[47m  [0m[40m  [0m[40m  [0m[40m  [0m[40m  [0m[40m  [0m[40m  [0m[40m  [0m[47m  [0m[47m  [0m[47m  [0m[47m  [0m[40m  [0m[40m  [0m[40m  [0m[47m  [0m[47m  [0m[40m  [0m[47m  [0m[40m  [0m[40m  [0m[40m  [0m[40m  [0m[40m  [0m[40m  [0m[40m  [0m[47m  [0m
[47m  [0m[40m  [0m[47m  [0m[47m  [0m[47m  [0m[47m  [0m[47m  [0m[40m  [0m[47m  [0m[47m  [0m[40m  [0m[40m  [0m[40m  [0m[47m  [0m[47m  [0m[40m  [0m[40m  [0m[40m  [0m[47m  [0m[40m  [0m[47m  [0m[47m  [0m[47m  [0m[47m  [0m[47m  [0m[40m  [0m[47m  [0m
[47m  [0m[40m  [0m[47m  [0m[40m  [0m[40m  [0m[40m  [0m[47m  [0m[40m  [0m[47m  [0m[40m  [0m[40m  [0m[47m  [0m[40m  [0m[47m  [0m[47m  [0m[40m  [0m[47m  [0m[40m  [0m[47m  [0m[40m  [0m[47m  [0m[40m  [0m[40m  [0m[40m  [0m[47m  [0m[40m  [0m[47m  [0m
[47m  [0m[40m  [0m[47m  [0m[40m  [0m[40m  [0m[40m  [0m[47m  [0m[40m  [0m[47m  [0m[40m  [0m[47m  [0m[40m  [0m[40m  [0m[40m  [0m[47m  [0m[47m  [0m[47m  [0m[47m  [0m[47m  [0m[40m  [0m[47m  [0m[40m  [0m[40m  [0m[40m  [0m[47m  [0m[40m  [0m[47m  [0m
[47m  [0m[40m  [0m[47m  [0m[40m  [0m[40m  [0m[40m  [0m[47m  [0m[40m  [0m[47m  [0m[40m  [0m[47m  [0m[47m  [0m[40m  [0m[40m  [0m[40m  [0m[47m  [0m[47m  [0m[40m  [0m[47m  [0m[40m  [0m[47m  [0m[40m  [0m[40m  [0m[40m  [0m[47m  [0m[40m  [0m[47m  [0m
[47m  [0m[40m  [0m[47m  [0m[47m  [0m[47m  [0m[47m  [0m[47m  [0m[40m  [0m[47m  [0m[40m  [0m[40m  [0m[40m  [0m[40m  [0m[47m  [0m[40m  [0m[40m  [0m[40m  [0m[40m  [0m[47m  [0m[40m  [0m[47m  [0m[47m  [0m[47m  [0m[47m  [0m[47m  [0m[40m  [0m[47m  [0m
[47m  [0m[40m  [0m[40m  [0m[40m  [0m[40m  [0m[40m  [0m[40m  [0m[40m  [0m[47m  [0m[40m  [0m[47m  [0m[40m  [0m[47m  [0m[40m  [0m[47m  [0m[40m  [0m[47m  [0m[40m  [0m[47m  [0m[40m  [0m[40m  [0m[40m  [0m[40m  [0m[40m  [0m[40m  [0m[40m  [0m[47m  [0m
[47m  [0m[47m  [0m[47m  [0m[47m  [0m[47m  [0m[47m  [0m[47m  [0m[47m  [0m[47m  [0m[40m  [0m[40m  [0m[40m  [0m[40m  [0m[47m  [0m[40m  [0m[47m  [0m[47m  [0m[47m  [0m[47m  [0m[47m  [0m[47m  [0m[47m  [0m[47m  [0m[47m  [0m[47m  [0m[47m  [0m[47m  [0m
[47m  [0m[40m  [0m[47m  [0m[40m  [0m[40m  [0m[40m  [0m[40m  [0m[40m  [0m[47m  [0m[47m  [0m[47m  [0m[47m  [0m[47m  [0m[47m  [0m[47m  [0m[40m  [0m[40m  [0m[47m  [0m[47m  [0m[40m  [0m[40m  [0m[40m  [0m[40m  [0m[40m  [0m[47m  [0m[47m  [0m[47m  [0m
[47m  [0m[47m  [0m[40m  [0m[40m  [0m[47m  [0m[47m  [0m[47m  [0m[47m  [0m[40m  [0m[47m  [0m[40m  [0m[40m  [0m[47m  [0m[47m  [0m[47m  [0m[47m  [0m[47m  [0m[40m  [0m[47m  [0m[47m  [0m[40m  [0m[47m  [0m[47m  [0m[47m  [0m[40m  [0m[47m  [0m[47m  [0m
[47m  [0m[47m  [0m[40m  [0m[40m  [0m[40m  [0m[47m  [0m[47m  [0m[40m  [0m[47m  [0m[40m  [0m[40m  [0m[40m  [0m[47m  [0m[47m  [0m[47m  [0m[47m  [0m[40m  [0m[47m  [0m[40m  [0m[47m  [0m[40m  [0m[40m  [0m[40m  [0m[47m  [0m[40m  [0m[40m  [0m[47m  [0m
[47m  [0m[47m  [0m[47m  [0m[40m  [0m[47m  [0m[40m  [0m[40m  [0m[47m  [0m[47m  [0m[40m  [0m[40m  [0m[40m  [0m[40m  [0m[40m  [0m[47m  [0m[40m  [0m[47m  [0m[40m  [0m[40m  [0m[47m  [0m[40m  [0m[47m  [0m[47m  [0m[47m  [0m[47m  [0m[40m  [0m[47m  [0m
[47m  [0m[40m  [0m[40m  [0m[47m  [0m[40m  [0m[40m  [0m[40m  [0m[40m  [0m[47m  [0m[40m  [0m[47m  [0m[40m  [0m[47m  [0m[47m  [0m[47m  [0m[47m  [0m[47m  [0m[47m  [0m[40m  [0m[40m  [0m[47m  [0m[40m  [0m[47m  [0m[40m  [0m[40m  [0m[40m  [0m[47m  [0m
[47m  [0m[40m  [0m[47m  [0m[40m  [0m[47m  [0m[40m  [0m[40m  [0m[47m  [0m[40m  [0m[47m  [0m[47m  [0m[40m  [0m[47m  [0m[40m  [0m[40m  [0m[40m  [0m[47m  [0m[47m  [0m[47m  [0m[47m  [0m[40m  [0m[47m  [0m[40m  [0m[47m  [0m[40m  [0m[47m  [0m[47m  [0m
[47m  [0m[40m  [0m[47m  [0m[47m  [0m[47m  [0m[40m  [0m[40m  [0m[40m  [0m[47m  [0m[40m  [0m[40m  [0m[40m  [0m[47m  [0m[47m  [0m[40m  [0m[40m  [0m[40m  [0m[40m  [0m[40m  [0m[47m  [0m[40m  [0m[40m  [0m[40m  [0m[47m  [0m[40m  [0m[40m  [0m[47m  [0m
[47m  [0m[40m  [0m[47m  [0m[47m  [0m[40m  [0m[47m  [0m[40m  [0m[47m  [0m[40m  [0m[47m  [0m[47m  [0m[40m  [0m[47m  [0m[40m  [0m[40m  [0m[47m  [0m[40m  [0m[40m  [0m[40m  [0m[47m  [0m[40m  [0m[40m  [0m[47m  [0m[47m  [0m[47m  [0m[40m  [0m[47m  [0m
[47m  [0m[40m  [0m[47m  [0m[40m  [0m[47m  [0m[40m  [0m[40m  [0m[40m  [0m[47m  [0m[40m  [0m[40m  [0m[40m  [0m[47m  [0m[40m  [0m[47m  [0m[40m  [0m[40m  [0m[40m  [0m[40m  [0m[40m  [0m[40m  [0m[40m  [0m[47m  [0m[40m  [0m[47m  [0m[47m  [0m[47m  [0m
[47m  [0m[47m  [0m[47m  [0m[47m  [0m[47m  [0m[47m  [0m[47m  [0m[47m  [0m[47m  [0m[40m  [0m[40m  [0m[40m  [0m[47m  [0m[47m  [0m[47m  [0m[47m  [0m[40m  [0m[40m  [0m[47m  [0m[47m  [0m[47m  [0m[40m  [0m[40m  [0m[47m  [0m[47m  [0m[47m  [0m[47m  [0m
[47m  [0m[40m  [0m[40m  [0m[40m  [0m[40m  [0m[40m  [0m[40m  [0m[40m  [0m[47m  [0m[47m  [0m[47m  [0m[47m  [0m[40m  [0m[40m  [0m[47m  [0m[47m  [0m[47m  [0m[40m  [0m[47m  [0m[40m  [0m[47m  [0m[40m  [0m[47m  [0m[40m  [0m[40m  [0m[40m  [0m[47m  [0m
[47m  [0m[40m  [0m[47m  [0m[47m  [0m[47m  [0m[47m  [0m[47m  [0m[40m  [0m[47m  [0m[40m  [0m[47m  [0m[47m  [0m[40m  [0m[47m  [0m[47m  [0m[40m  [0m[47m  [0m[40m  [0m[47m  [0m[47m  [0m[47m  [0m[40m  [0m[40m  [0m[47m  [0m[47m  [0m[47m  [0m[47m  [0m
[47m  [0m[40m  [0m[47m  [0m[40m  [0m[40m  [0m[40m  [0m[47m  [0m[40m  [0m[47m  [0m[40m  [0m[47m  [0m[47m  [0m[40m  [0m[47m  [0m[47m  [0m[47m  [0m[40m  [0m[40m  [0m[40m  [0m[40m  [0m[40m  [0m[40m  [0m[47m  [0m[40m  [0m[47m  [0m[40m  [0m[47m  [0m
[47m  [0m[40m  [0m[47m  [0m[40m  [0m[40m  [0m[40m  [0m[47m  [0m[40m  [0m[47m  [0m[40m  [0m[47m  [0m[40m  [0m[47m  [0m[40m  [0m[40m  [0m[40m  [0m[40m  [0m[47m  [0m[40m  [0m[40m  [0m[47m  [0m[40m  [0m[40m  [0m[40m  [0m[40m  [0m[40m  [0m[47m  [0m
[47m  [0m[40m  [0m[47m  [0m[40m  [0m[40m  [0m[40m  [0m[47m  [0m[40m  [0m[47m  [0m[40m  [0m[40m  [0m[47m  [0m[47m  [0m[47m  [0m[40m  [0m[40m  [0m[47m  [0m[40m  [0m[47m  [0m[47m  [0m[47m  [0m[47m  [0m[40m  [0m[40m  [0m[47m  [0m[40m  [0m[47m  [0m
[47m  [0m[40m  [0m[47m  [0m[47m  [0m[47m  [0m[47m  [0m[47m  [0m[40m  [0m[47m  [0m[47m  [0m[47m  [0m[47m  [0m[47m  [0m[40m  [0m[40m  [0m[47m  [0m[40m  [0m[40m  [0m[40m  [0m[40m  [0m[40m  [0m[40m  [0m[40m  [0m[47m  [0m[47m  [0m[40m  [0m[47m  [0m
[47m  [0m[40m  [0m[40m  [0m[40m  [0m[40m  [0m[40m  [0m[40m  [0m[40m  [0m[47m  [0m[40m  [0m[40m  [0m[40m  [0m[47m  [0m[40m  [0m[47m  [0m[40m  [0m[47m  [0m[47m  [0m[40m  [0m[47m  [0m[40m  [0m[40m  [0m[40m  [0m[40m  [0m[40m  [0m[40m  [0m[47m  [0m
[47m  [0m[47m  [0m[47m  [0m[47m  [0m[47m  [0m[47m  [0m[47m  [0m[47m  [0m[47m  [0m[47m  [0m[47m  [0m[47m  [0m[47m  [0m[47m  [0m[47m  [0m[47m  [0m[47m  [0m[47m  [0m[47m  [0m[47m  [0m[47m  [0m[47m  [0m[47m  [0m[47m  [0m[47m  [0m[47m  [0m[47m  [0m
"
`;

exports[`utf8 should turn into a string 1`] = `
"                                 
                                 
    █▀▀▀▀▀█  ▄▄█▀▀▄▄█ █▀▀▀▀▀█    
    █ ███ █ █▀▄█▄ ▀ ▀ █ ███ █    
    █ ▀▀▀ █ █▄▄█▀█▄▄█ █ ▀▀▀ █    
    ▀▀▀▀▀▀▀ █▄█▄▀▄▀ ▀ ▀▀▀▀▀▀▀    
    ▀▄█▀▀▀▀▄ ▄▄   ▀▀▄ ▀█▀▀▀▄     
     ▀█▀▄▄▀ ███▄▄ ▄▀▄█ █▀▀ ▀█    
    █▀▄▀██▀▄▀ █ ▄▄▄  ▀▀▄▀▄▀█▀    
    █  ▄▀█▀▄▀▀█ ▄█▀███ ██▀ ▀█    
    ▀ ▀ ▀▀▀ ███ ▀ ▀██▀▀▀█▄▀      
    █▀▀▀▀▀█ ▄  █▀ ▄ █ ▀ █▄▀▀▀    
    █ ███ █ █ ▄▀▄▄▄█▀██▀█▄█▄█    
    █ ▀▀▀ █ ▀▀  ▄█▀▄█▄▄▄▄█▀ █    
    ▀▀▀▀▀▀▀ ▀▀▀ ▀ ▀  ▀ ▀▀▀▀▀▀    
                                 
                                 "
`;
